src/Microsoft.OpenApi/Services/OpenApiWalker.cs

```diff
@@ -22,81 +24,134 @@ public OpenApiWalker(OpenApiVisitorBase visitor)
         }
 
         /// <summary>
-        /// Walks through the <see cref="OpenApiDocument"/> and validates each element.
+        /// Walks through the <see cref="OpenApiDocument"/> and visits each element.
         /// </summary>
-        /// <param name="doc"></param>
+        /// <param name="doc">OpenApiDocument to be walked</param>
         public void Walk(OpenApiDocument doc)
         {
             _visitor.Visit(doc);
             _visitor.Visit(doc.Info);
             _visitor.Visit(doc.Servers);
 
+            // Visit Servers
             if (doc.Servers != null)
             {
                 foreach (var server in doc.Servers)
                 {
-                    _visitor.Visit(server);
-                    foreach (var variable in server.Variables.Values)
-                    {
-                        _visitor.Visit(variable);
-                    }
+                    WalkServer(server);
                 }
             }
 
+            // Visit Paths
             _visitor.Visit(doc.Paths);
             foreach (var pathItem in doc.Paths.Values)
             {
-                _visitor.Visit(pathItem);
-                _visitor.Visit(pathItem.Operations);
-                foreach (var operation in pathItem.Operations.Values)
+                WalkPathItem(pathItem);
+            }
+
+            _visitor.Visit(doc.Components);
+
+            _visitor.Visit(doc.ExternalDocs);
+
+            _visitor.Visit(doc.Tags);
+
+            foreach (var tag in doc.Tags)
+            {
+                WalkTag(tag);
+            }
+
+            _visitor.Visit(doc as IOpenApiExtensible);
+
+        }
```

Comment:
It took me a bit to understand how to distinguish the case when we simply call `.Visit` and the case where we need to have `WalkXXX` private method.

Would it be easier to read if we always create WalkXXX method for every OpenApiXXX and its collection properties? If it's something simple without further inner object like OpenApiInfo, then sure, WalkInfo will simply be a wrapper around _visitor.Visit(info).

My concern is that with the current implementation, in case someone mistakenly uses .Visit instead of Walk in certain places, it will be very difficult to spot.

If we always create a WalkXXX, the pattern for each WalkXXX is also always consistent, with one Visit call and a bunch of following Walk calls.

```c#
public void WalkObject(OpenApiObject object)
{
    // Visit self
    _visitor.Visit(object);
    
     // Walk the rest
     WalkProperty1(object.Property1);
     WalkMapProperty2(object.MapProperty2);
}

public void WalkProperty1(OpenApiProperty1Type property1)
{
     // Visit self
     _visitor.Visit(property1);
     
     // Walk the rest
     WalkProperty11(property1.Property11);
     WalkProperty12(property1.Property12);
     WalkProperty13(property1.Property13);
}

public void WalkMapProperty2(IDictionary<string, Property2Type> property2Map)
{
     // Visit self
     _visitor.Visit(property2Map);
     
     // Walk the rest
     foreach( var property2 in property2Map.Values )
     {
           WalkProperty2(property2);
     }
}
```

This will also address Sam's concern about having other starting point for Walking.

Reply:
Walk methods are just helper methods for grouping visits.  Think of Walk methods as non-leaf nodes of the traversal tree and Visit as the leaf nodes.  If you want to have a Walk for every leaf I suppose we could.

Reply:
I think it's a good thing to have. Say, if customer wants to only walk his OpenApiComponents, having a WalkComponents will enable that feature.

---
In reply to: [158121152](https://github.com/Microsoft/OpenAPI.NET/pull/163#discussion_r158121152) [](ancestors = 158121152)

Reply:
You mean they only want to Validate the components, but not validate anything else?  Or do you mean they want to only apply a visitor to the components?  If so, they can create a class that derives from the VisitorBase, create an overload of the Visit(OpenApiComponent ) method and then use the regular walk.


---
In reply to: [158121661](https://github.com/Microsoft/OpenAPI.NET/pull/163#discussion_r158121661) [](ancestors = 158121661,158121152)

Reply:
I meant the former. Given that the implementation consistency concern is relatively minor, this boils down to the same issue as Sam's.

We can just discuss in the other thread in OpenApiWalker.cs to avoid repetition.

---
In reply to: [158126857](https://github.com/Microsoft/OpenAPI.NET/pull/163#discussion_r158126857) [](ancestors = 158126857,158121661,158121152)